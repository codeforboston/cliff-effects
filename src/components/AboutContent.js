import React from 'react';
import { Header, Message } from 'semantic-ui-react';
import { ExternalLink } from '../forms/formHelpers';

// CUSTOM COMPONENTS

/** Todo simplify the process of creating content for the 'About' page */

const AboutContent = function ({ snippets }) {
  return (
    <div>

      <Header
        as='h1'
        style={{ fontSize: '4em' }}>
        { snippets.aboutPageHeader_v1 }
      </Header>

      <Header as='h3'>{ snippets.whatForHeader_v1 }</Header>
      <Message>{ snippets.whatForImportantNote_v1 }</Message>
      <p>{ snippets.whatFor_v2 }</p>

      <Header as='h3'>{ snippets.whyHeader_v1 }</Header>
      <p>{ snippets.why1_v1 }</p>
      <p>{ snippets.why2_v1 }</p>

      <ul>
        <li>
          <ExternalLink href="https://www.youtube.com/watch?v=BveX_rID4_E">
            { snippets.videoLinkText_v1 }
          </ExternalLink>
        </li>

        <li>
          <ExternalLink href="http://www.nccp.org/projects/files/NCCP_CO_presentation07.pdf">
            { snippets.quantLinkText_v1 }
          </ExternalLink>
        </li>

        <li>
          <ExternalLink href="https://www.umb.edu/editor_uploads/images/centers_institutes/center_social_policy/The_Road_to_the_Cliff_Edge_08.16.17.pdf">
            { snippets.benefitsLinkText_v1 }
          </ExternalLink>
        </li>
      </ul>

      <Header as='h3'>{ snippets.howToUseHeader_v1 }</Header>
      <p>{ snippets.howToUse_v1 }</p>
      <Message>{ snippets.howToUseNote_v1 }</Message>

      <Header as='h3'>{ snippets.whoMadeThisHeader_v1 }</Header>
      <p>{ snippets.whoMadeThis1_v1 }</p>
      <p>{ snippets.whoMadeThis2_v1 }</p>
      <p>{ snippets.whoMadeThis3_v1 }</p>

    </div>
  );

};  // End AboutContent(<>)


export { AboutContent };
