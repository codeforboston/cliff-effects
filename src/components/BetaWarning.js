import React from 'react';
import { Menu, Responsive } from 'semantic-ui-react';

const BetaWarning = function (props) {
  return (
    <Menu.Item>
      <Responsive
        as='strong'
        {...Responsive.onlyTablet}>
        This tool is a prototype.
      </Responsive>
      <Responsive
        as='strong'
        {...Responsive.onlyComputer}>
        This tool is a prototype and should not be used to make financial decisions.
      </Responsive>
    </Menu.Item>
  );
};  // End BetaWarning(<>)


export { BetaWarning };
