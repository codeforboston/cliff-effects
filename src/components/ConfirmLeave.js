import React from 'react';

/**
 * Ask the user for confirmation to leave the page.
 * @extends React.Component
 *
 * @param props {object}
 * @param props.isBlocking {boolean} - Whether the component should block
 * @param props.message {string} - The message the browser shows, maybe
 */
class ConfirmLeave extends React.Component {
  static defaultProps = { isBlocking: true };

  confirm = (event) => {
    if (!this.props.isBlocking) {
      return; // do not block unload
    } 
    return (event.returnValue = this.props.message || '');
  };

  componentDidMount() {
    window.addEventListener('beforeunload', this.confirm);
  }

  componentWillUnmount() {
    window.removeEventListener('beforeunload', this.confirm);
  }

  render() {
    return null;
  }
}

export default ConfirmLeave;
