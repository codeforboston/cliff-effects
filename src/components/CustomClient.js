import React from 'react';
import { Divider, Form, Message } from 'semantic-ui-react';

/**
 * Load previous session from object.
 * @callback loadClient
 * @param {object} clientContainer - Object containing previous session as `client`.
 */

/**
 * Form which loads previous session from JSON.
 * 
 * @param {object} props
 * @param {boolean} props.mayLoadCustomClient - Whether form should be visible
 * @param {loadClient} props.loadClient
 * 
 * @extends React.Component
 */
class CustomClient extends React.Component {
  state = {
    client: null,
    error:  null,
    json:   '',
  };

  submit = (event) => {
    const { client } = this.state;
    event.preventDefault();
    if (client === null) {
      return;
    }

    this.setState({
      client: null,
      error:  null,
      json:   '',
    });
    this.props.loadClient({ client: client });
  };

  handleChange = (_event, inputProps) => {
    const { value } = inputProps;
    try {
      const newClient = JSON.parse(value);
      this.setState({
        client: newClient,
        error:  null,
        json:   value,
      });
    } catch (error) {
      this.setState({
        client: null,
        error:  error,
        json:   value,
      });
    }
  };

  render() {
    if (!this.props.mayLoadCustomClient) {
      return null;
    }
    const { client, error, json } = this.state;

    return (
      <Form
        error={ error !== null }
        onSubmit={ this.submit }>
        <Form.Field>
          <label>Client JSON</label>
          <Form.Input
            type={ 'text' }
            value={ json }
            onChange={ this.handleChange } />
        </Form.Field>
        <Message 
          error
          header={ 'JSON Parse Failed!' }
          content={ error && error.message } />
        <Form.Button
          type={ 'submit' }
          disabled={ client === null }>
          Import Data
        </Form.Button>
        <Divider />
      </Form>
    );
  }
}

export { CustomClient };
