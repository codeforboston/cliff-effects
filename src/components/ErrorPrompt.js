import React from 'react';

/**
 * Prompt for latest uncaught error.
 * @extends React.Component
 *
 * @param props {object}
 * @param props.callback {onLeaveCallback}
 * @param props.message {string}
 * @param props.prompt {prompt}
 */
class ErrorPrompt extends React.Component {
  handleError = ({ error }) => {
    const { callback, prompt, ...promptProps } = this.props;
    prompt(callback, promptProps);
  };

  componentDidMount() {
    window.addEventListener('error', this.handleError);
  }

  componentWillUnmount() {
    window.removeEventListener('error', this.handleError);
  }

  render() {
    return null;
  }
}

export default ErrorPrompt;
