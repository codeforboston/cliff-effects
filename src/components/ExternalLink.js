import React from 'react';
/**
 * Link that opens new tab
 */
const ExternalLink = function ({ href, children, ...otherProps }) {
  return (
    <a
      href   = { href }
      target = { `_blank` }
      { ...otherProps }>
      {children}
    </a>
  );
};

export { ExternalLink };
