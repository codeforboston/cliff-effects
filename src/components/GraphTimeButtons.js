import { Button } from 'semantic-ui-react';
import React from 'react';

const GraphButton = function ({ id, activeID, onClick }) {
  return (
    <Button
      id={id}
      active={activeID === id}
      onClick={onClick}>
      {id}
    </Button>
  );
};

const GraphTimeButtons = function ({ activeID, onClick }) {
  return (
    <Button.Group
      basic
      className='graph-time-options'>
      <GraphButton
        id={'Weekly'}
        activeID={activeID}
        onClick={onClick} />
      <GraphButton
        id={'Monthly'}
        activeID={activeID}
        onClick={onClick} />
      <GraphButton
        id={'Yearly'}
        activeID={activeID}
        onClick={onClick} />
    </Button.Group>
  );
};

export { GraphTimeButtons };
