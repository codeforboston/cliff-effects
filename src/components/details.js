import React, { Component } from 'react';
import {
  Icon,
  Transition,
  Message,
} from 'semantic-ui-react';


/** Really it's a top component that has a question mark
 *    next to it followed by a bottom one that starts
 *    as hidden, but is revealed when the question mark
 *    is clicked.
 *
 * @param {object} props
 * @param {object} props.children Must have exactly two
 *    children. The first is always visible. The second
 *    is revealed when the icon is clicked and then hidden
 *    when the icon is clicked again.
 */
class HeadingWithDetail extends Component {
  state = { showDetails: false };

  toggleDetails = () => {
    this.setState((prevState) => {
      return { showDetails: !prevState.showDetails };
    });
  };

  render () {
    var { showDetails } = this.state;

    var children  = this.props.children;
    if (!Array.isArray(children) || children.length !== 2) {
      console.warn(`<HeadingWithDetail> must have exactly two children`);
      return null;
    }

    var top       = children[ 0 ] || children,
        details   = children[ 1 ] || null,
        // Has a green question mark icon by default
        iconName  = `question circle outline`,
        iconColor = `teal`;

    // Show a red 'x' instead if the detail is visible right now
    if (showDetails) {
      iconName  = `remove circle`;
      iconColor = `red`;
    }

    return (
      <div className = { `heading-with-detail` }>
        <div className={ `has-details` }>{ top }</div>
        <Icon
          name      = { iconName }
          color     = { iconColor }
          className = { `details-icon` }
          onClick   = { this.toggleDetails } />
        <Transition
          visible   = { showDetails }
          animation = { `slide down` }>
          <Message>{ details }</Message>
        </Transition>
      </div>
    );
  };
};  // End <HeadingWithDetail>


/** @todo Have rows that can have details. Maybe other things too. */
class RowWithDetail extends Component {};  // <End RowWithDetail>


export {
  HeadingWithDetail,
  RowWithDetail,
};
