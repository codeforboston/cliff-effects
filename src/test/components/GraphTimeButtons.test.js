import React from 'react';
import { mount } from 'enzyme';

import { GraphTimeButtons } from '../../components/GraphTimeButtons';

describe('<GraphTimeButtons>', () => {
  it('renders Weekly, Monthly, and Yearly buttons', () => {
    const wrapper = mount(<GraphTimeButtons />);
    expect(wrapper.find('#Weekly').exists()).toBe(true);
    expect(wrapper.find('#Monthly').exists()).toBe(true);
    expect(wrapper.find('#Yearly').exists()).toBe(true);
  });

  it('renders with provided props', () => {
    const props = {
      activeID:   'Monthly',
      onClick:    jest.fn(),
      otherProps: 'I don\'t get passed through =(',
    };
    expect(mount(<GraphTimeButtons { ...props } />)).toMatchSnapshot();
  });
});
