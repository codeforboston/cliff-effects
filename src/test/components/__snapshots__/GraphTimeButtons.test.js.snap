// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GraphTimeButtons> renders with provided props 1`] = `
<GraphTimeButtons
  activeID="Monthly"
  onClick={[Function]}
  otherProps="I don't get passed through =("
>
  <ButtonGroup
    basic={true}
    className="graph-time-options"
  >
    <div
      className="ui basic buttons graph-time-options"
    >
      <GraphButton
        activeID="Monthly"
        id="Weekly"
        onClick={[Function]}
      >
        <Button
          active={false}
          as="button"
          id="Weekly"
          onClick={[Function]}
        >
          <button
            className="ui button"
            id="Weekly"
            onClick={[Function]}
            role="button"
          >
            Weekly
          </button>
        </Button>
      </GraphButton>
      <GraphButton
        activeID="Monthly"
        id="Monthly"
        onClick={[Function]}
      >
        <Button
          active={true}
          as="button"
          id="Monthly"
          onClick={[Function]}
        >
          <button
            className="ui active button"
            id="Monthly"
            onClick={[Function]}
            role="button"
          >
            Monthly
          </button>
        </Button>
      </GraphButton>
      <GraphButton
        activeID="Monthly"
        id="Yearly"
        onClick={[Function]}
      >
        <Button
          active={false}
          as="button"
          id="Yearly"
          onClick={[Function]}
        >
          <button
            className="ui button"
            id="Yearly"
            onClick={[Function]}
            role="button"
          >
            Yearly
          </button>
        </Button>
      </GraphButton>
    </div>
  </ButtonGroup>
</GraphTimeButtons>
`;
